### ISSUE TITLE
[Bug] Android App Links deeplink blocked (Custom Tabs)

### ISSUE THREAD

**Issue by lbalmaceda** [mozilla-mobile/fenix#15202]
Hi team, 👋 
I'm using [Android App Links](https://developer.android.com/training/app-links/); the official way of doing deep linking since Android API 23. In that scenario, the domains are previously verified by Android at -user application- install time and their intent-filters are approved or not. Once verified, visiting a link that matches those intent-filter criteria, would invoke their app without asking the user for choosing one app.

When visiting a callback URL such as "https://myapp.com/callback" the Chrome browser redirects successfully without further action. On Firefox, that's not the case. Because the URL doesn't have any content, as it's meant to be used only for registering that redirect, Firefox displays a "Not found." text and white page in the custom tab, and unless I select "Open in app" from the menu, I won't be redirected to the app.

![image](https://user-images.githubusercontent.com/3900123/93645374-3f651f80-f9da-11ea-9822-d49c800cadfe.png)

The question for you is, **how is this menu option intuitive for someone that just installs the Firefox browser?**. 

We have customers that are losing conversion rates because of this design decision. Their users don't understand how to log in to their applications after they come across this white page. And I couldn't find any official documentation on it that I could share with them. 

## Steps to reproduce
1. Create an android application targeting as min API android version 23. 
2. Register an intent filter that matches against a callback URL you can redirect to from your site.
3. Set `autoVerify=true` so the domain verification happens at install time.
4. Host the assets links file in the well-known directory of your domain, using the hash for the key you are using to sign the APK.
5. Install a release in the device, making sure it was signed with the same key from the previous step. (All this is detailed in the android docs linked at the top).
6. Run your app, call a custom tab that opens a site with a button that when clicked would invoke that callback URL you've registered in the intent-filter.
7. Watch how firefox ignores the redirect when you click the button.
8. Click the "Open in app" menu option and watch how it redirects successfully.

I'm testing this on a sample app using the [Auth0 SDK](https://github.com/auth0/Auth0.Android).

### Expected behavior
Redirect back to the app automatically, **without** additional user action.

### Actual behavior
The page shows content loading error "Not found." and the user has to click the overflow menu `...` and select "Open in app" for the app to successfully receive the intent.

What I'd expect from Firefox is to at least check those apps that have officially registered their intent-filters using App Links and not thinking twice about opening it on that same browser tab or in an external app. The domain has been already verified by the OS at install time using the hosted `.well-known/assetslinks.json` file, so the deep link should be trusted. 

I've also tested first enabling the "Open links in apps" but that didn't change anything, at least when used from a Custom Tab.

![image](https://user-images.githubusercontent.com/3900123/93646190-1e9dc980-f9dc-11ea-8fc1-8edc9437f630.png)

### Device information
I tried this on a Pixel 2 emulator with Google Play Services, running Android 28. I've installed the latest Firefox app from the [Playstore at that time](https://play.google.com/store/apps/details?id=org.mozilla.firefox&hl=en) (yesterday). I don't think that's the "nightly" version, but rather the classic. The version name according to the Playstore listing is `80.1.3`.

* Android device: Pixel 2, with Google Apps installed. Android API 28.
* Fenix version: ?

Related links:
- https://support.mozilla.org/en-US/kb/set-firefox-android-open-links-native-apps
- https://support.mozilla.org/en-US/questions/1297628

**Comment by lbalmaceda** [comment]
Hi team, any updates on this matter?

**Comment by OlegAlekos** [comment]
More and more apps want to take advantage of the App-Link feature. 
Every browser with chromium supports this feature, it requires a lot of custom handling and workarounds just to handle this case of FireFox. 
Any updates ?

**Comment by carstenhag** [comment]
I'm pretty sure Firefox has to change this, right? We stumbled across this firefox feature/bug and are now pretty stuck.  

Pretty sure the only way all of our users will be able to log in (they don't understand the blue dot) is by denylisting Firefox at our login library's configuration. See also https://github.com/openid/AppAuth-Android/issues/732.

**Comment by rocketsroger** [comment]
@lbalmaceda I can help you with this issue.  From your description, I'm not seeing a reason why it wouldn't open the link in your app.  Is there a link + app combination I can use to test your specific case?  Thanks,

**Comment by rocketsroger** [comment]
@carstenhag your issue seems to be the same, Is there a way I can test your case as well?  thanks

**Comment by carstenhag** [comment]
Hello! Yes, I can provide you an APK + test account, where can I privately send you this?

**Comment by rocketsroger** [comment]
> Hello @rocketsroger! Yes, I can provide you an APK + test account, where can I privately send you this? Mozilla's chat?

Yes, that'll work for me.  Thanks

**Comment by carstenhag** [comment]
Info for the others here: @rocketsroger and I had a chat and he looked at our app & the login. A summary of the chat:

>He agrees that the current UX is bad. While there is a settings option to open the link in the app, many users will not even understand the option. It is currently implemented as is to protect the user's privacy.
He says when using Custom Tabs, perhaps this setting could be ignored, and a redirect would always open a native app.
He will talk to the "product people" to see what can be done. He thanked me for being vocal about this.

**Comment by rocketsroger** [comment]
Pinning this to keep this on my to do list.

**Comment by GNUGradyn** [comment]
Any update on this? This completely breaks any links that are sent to the user that should open in the app when the users default browser is Firefox and there is any sort of "middle man link" between the deep link and the link sent to the user. **This includes firebase dynamic links. Apps that use firebase dynamic links are currently broken for firefox users**. This should be a bigger deal then it is. Here is a very common example configuration that is broken for all users with firefox as their default browser

1. A Firebase link is sent to the customer via text or email or whatever
2. The user clicks the firebase link. The firebase link is opened in firefox
3. Firebase determines the user has the app and is running android and therefore redirects to the deep link
4. We are now on the deep link, but we are still in firefox. This generally results in a 404 being displayed when the user should have been sent to the app

Currently this pretty much forces developers to either not use firebase dynamic links or anything similar, or blacklist firefox entirely. FIrefox should be handling deep links like chromium based browsers

**Comment by adfleshner** [comment]
Any news on if this is going to be updated or changed. I came across this because I am testing Google Custom Tabs https://developer.chrome.com/docs/android/custom-tabs/ and applinks and have run into a snag that if FireFox is the default browser the redirect back into the app does not work. I realize that there is a "Open in app" button but it is hidden and the user may not realize that is the reason why they can not log in. Is there any movement/timeline on a fix for this?

**Comment by carstenhag** [comment]
@adfleshner there is a Firefox setting that no user will know about, that changes the behavior :P. If you are using appauth or a similar login lib, the only good solution is to forbid Firefox.  
In any other case, you are out of luck at the moment.

**Comment by CholoTook** [comment]
What's the simplest workaround for a dumb dev like me? Tell users to remove firefox?

**Comment by rocketsroger** [comment]
> Any news on if this is going to be updated or changed. I came across this because I am testing Google Custom Tabs https://developer.chrome.com/docs/android/custom-tabs/ and applinks and have run into a snag that if FireFox is the default browser the redirect back into the app does not work. I realize that there is a "Open in app" button but it is hidden and the user may not realize that is the reason why they can not log in. Is there any movement/timeline on a fix for this?

Apologize for not fixing this already.  I spent some time looking for a simple solution a while ago but no luck.  I'll allocate some more time soon to fix this.  Sorry again for not getting to this sooner.

**Comment by GNUGradyn** [comment]
For anyone trying to work around this, you can work around it in some scenarios by making sure the user *directly* receives a deep link. Redirecting to a deep link will not work.

**Comment by CholoTook** [comment]
> For anyone trying to work around this, you can work around it in some scenarios by making sure the user _directly_ receives a deep link. Redirecting to a deep link will not work.

So, for example, instead of detecting the device and redirecting to the app, there would be a button saying 'open in the app' that directly deep linked? Is that working in FF?

**Comment by rocketsroger** [comment]
Looking into this more, the main issue here is that most users don't understand "Open links in apps" means.  Since this is default to false, most of the users don't turn it on.  I think there are two ways to fix this:

1. Turn "Open links in apps" default on:
- This won't help users that have already turned it off.
- Maybe users only want some links to redirect to 3rd party apps but this switch is all or none.

2. Remove the switch and always prompt the user if they want to redirect to a 3rd party app. (Similar to what we do in private tabs right now)
- User will need an extra tap before getting redirect to a 3rd party app.

**Comment by GNUGradyn** [comment]
Perhaps make it 3 options:
"Never open links in third party apps"
"Prompt every time"
"Always open links in third party apps"

And make the second one default?

**Comment by rocketsroger** [comment]
> Perhaps make it 3 options: "Never open links in third party apps" "Prompt every time" "Always open links in third party apps"
> 
> And make the second one default?

This could be a solution as well.  Thanks

**Comment by carstenhag** [comment]
@rocketsroger Prompting would be fine for me. 
But perhaps it would be good to challenge this setting altogether. It comes from a time where any 3rd party developer could "listen" for https://wikipedia.com links and the Android OS would show a picker to allow opening that link in a 3rd party app.

But since Android 12 (I couldn't quickly find a good link to android.com, but here's an article https://doordash.engineering/2022/01/25/your-deep-links-might-be-broken-web-intents-and-android-12/) only 1st party (verified app-domain combinations) can get opened when clicking on an URL. 

To sum up: Since Android 6 all apps could be opened with a generic https URL, with android 12 only specific verified apps can be opened with https App Links.

**Comment by rocketsroger** [comment]
I did not get confirmation yet.  I will continue next week to push this issue forward. Thanks

**Comment by GNUGradyn** [comment]
@carstenhag makes a good point. Perhaps the third option I listed could be default on 12+ and the middle one default on 11-

**Comment by rileyinman** [comment]
Notably, apps can still be manually associated with domains: https://developer.android.com/about/versions/12/behavior-changes-all#web-intent-resolution

That said, as far as I can tell, it is true that only one app can be associated with a given domain.

**Comment by angaaron** [comment]
@rocketsroger what sort of timelines are you looking at for this? It's impacting a few of our customers at the moment so we will most likely have to implement a solution to remove firefox from opening the links.

**Comment by rocketsroger** [comment]
> @rocketsroger what sort of timelines are you looking at for this? It's impacting a few of our customers at the moment so we will most likely have to implement a solution to remove firefox from opening the links.

@angaaron I'm currently narrowing down what is the best change we can do here.  From the interactions on this issue, we know this is a pain point and we want to fix it quickly but correctly.  The question here is privacy and control vs. simplicity and convenience.  I am hoping to find a solution that can satisfy both sides without breaking any current use cases.  I'm hoping to resolve this in a month or so.  Thanks

**Comment by GNUGradyn** [comment]
Is there still a privacy concern in Android 12+? In Android 12+ the web page and the app have to mutually agree for the app to handle the link

**Comment by rocketsroger** [comment]
> Is there still a privacy concern in Android 12+? In Android 12+ the web page and the app have to mutually agree for the app to handle the link

Yes,  assuming you're searching for something and it redirects to a page on Wikipedia.  If by default we redirect to the Wikipedia app (if installed) right away, Firefox would have disclosed that search information to a 3rd party app without the user's consent/control.  I understand some apps it makes sense (Google maps, Wikipedia etc); however, users might want control on some other apps so they can have control over their data.

**Comment by carstenhag** [comment]
No, see my comment from 5 days ago. A Wikipedia link can only be opened automatically by the Wikipedia app (or a 3rd party app that Wikipedia grants access to do this).

But yes I see the point from Firefox to not opening the  app even if it's a 1st party one.

**Comment by rocketsroger** [comment]
> But yes I see the point from Firefox to not opening the app even if it's a 1st party one.

Agreed, even if the app is 1st party app we should still give user the control.  (Sorry, I should have said 1st party apps)

**Comment by csadilek** [comment]
Moved to bugzilla: https://bugzilla.mozilla.org/show_bug.cgi?id=1809264

Change performed by the [Move to Bugzilla add-on](https://addons.mozilla.org/en-US/firefox/addon/move-to-bugzilla/).
